extensions/amp-a4a/0.1/legacy-signature-verifier.js

```diff
@@ -14,51 +14,300 @@
+          }));
+    } else {
+      // The given signingServiceName does not have a corresponding URL in
+      // _a4a-config.js.
+      const reason = `Signing service '${signingServiceName}' does not exist.`;
+      user().error(TAG, reason);
+      this.keys_.push(/** @type {!Promise<!CryptoKeysDef>} */ (Promise.resolve({
+        signingServiceName,
+        keys: [],
+      })));
+    }
+  }
+
+  /**
+   * Extracts a cryptographic signature from `headers` and verifies that it's
+   * the correct cryptographic signature for `creative`.
+   *
+   * As a precondition, `loadKeyset` must have already been called on the
+   * signing service that was used.
+   *
+   * @param {!ArrayBuffer} creative
+   * @param {!Headers} headers
+   * @param {function(string, !Object)} lifecycleCallback called for each AMP
+   *     lifecycle event triggered during verification
+   * @return {!Promise<?VerificationFailure>} resolves to `null` on success, or,
+   *     on failure, to a `VerificationFailure` indicating the cause
+   */
+  verify(creative, headers, lifecycleCallback) {
+    if (!this.isAvailable_()) {
+      return Promise.resolve(
+          /** @type {?VerificationFailure} */ (VerificationFailure.NO_FAULT));
+    }
+    const headerValue = headers.get(AMP_SIGNATURE_HEADER);
+    if (!headerValue) {
+      return Promise.resolve(
+          /** @type {?VerificationFailure} */ (VerificationFailure.NO_FAULT));
+    }
+    let signature;
+    try {
+      signature = base64UrlDecodeToBytes(headerValue);
+    } catch (e) {
+      return Promise.resolve(/** @type {?VerificationFailure} */ (
+          VerificationFailure.SIGNATURE_MISMATCH));
+    }
+    // For each signing service, we have exactly one Promise,
+    // keyInfoSetPromise, that holds an Array of Promises of signing keys.
+    // So long as any one of these signing services can verify the
+    // signature, then the creative is valid AMP.
+    // Track if verification found, as it will ensure that promises yet to
+    // resolve will "cancel" as soon as possible saving unnecessary resource
+    // allocation.
+    let verified = false;
+    return some(this.keys_.map(keyInfoSetPromise => {
+      // Resolve Promise into an object containing a 'keys' field, which
+      // is an Array of Promises of signing keys.  *whew*
+      return keyInfoSetPromise.then(keyInfoSet => {
+        // As long as any one individual key of a particular signing
+        // service, keyInfoPromise, can verify the signature, then the
+        // creative is valid AMP.
+        if (verified) {
+          return Promise.reject('redundant');
+        }
+        return some(keyInfoSet.keys.map(keyInfoPromise => {
+          // Resolve Promise into signing key.
+          return keyInfoPromise.then(keyInfo => {
+            if (verified) {
+              return Promise.reject('redundant');
+            }
+            if (!keyInfo) {
+              return Promise.reject('Promise resolved to null key.');
```

Comment:
This would happen if the JSON object from the signing service could not be successfully converted into a crypto key. We do log to `user().error` when that happens, but at import time, not at verification time.